diff --git a/fs/read_write.c b/fs/read_write.c
--- a/fs/read_write.c
+++ b/fs/read_write.c
@@ -583,2 +583,6 @@ EXPORT_SYMBOL(vfs_write);
 
+#if defined(CONFIG_KSU_MANUAL_HOOK) && !defined(CONFIG_KSU_SUSFS)
+extern void ksu_handle_sys_read(unsigned int fd);
+#endif
+
 static inline loff_t file_pos_read(struct file *file)
@@ -599,2 +603,6 @@ SYSCALL_DEFINE3(read, unsigned int, fd, char __user *, buf, size_t, count)
 	ssize_t ret = -EBADF;
 
+#if defined(CONFIG_KSU_MANUAL_HOOK) && !defined(CONFIG_KSU_SUSFS)
+	ksu_handle_sys_read(fd);
+#endif
+
 	if (f.file) {
